src/expr/src/relation/func.rs

```diff
@@ -453,48 +458,98 @@ where
+fn jsonb_agg<'a, I>(datums: I, temp_storage: &'a RowArena, order_by: &Vec<ColumnOrder>) -> Datum<'a>
 where
     I: IntoIterator<Item = Datum<'a>>,
 {
+    let datums = order_aggregate_datums(datums, order_by);
     temp_storage.make_datum(|packer| {
         packer.push_list(datums.into_iter().filter(|d| !d.is_null()));
     })
 }
 
-fn jsonb_object_agg<'a, I>(datums: I, temp_storage: &'a RowArena) -> Datum<'a>
+fn jsonb_object_agg<'a, I>(
+    datums: I,
+    temp_storage: &'a RowArena,
+    order_by: &Vec<ColumnOrder>,
+) -> Datum<'a>
 where
     I: IntoIterator<Item = Datum<'a>>,
 {
+    let datums = order_aggregate_datums(datums, order_by);
     temp_storage.make_datum(|packer| {
-        packer.push_dict(
-            datums
-                .into_iter()
-                .filter_map(|d| {
-                    if d.is_null() {
-                        return None;
-                    }
-                    let mut list = d.unwrap_list().iter();
-                    let key = list.next().unwrap();
-                    let val = list.next().unwrap();
-                    if key.is_null() {
-                        // TODO(benesch): this should produce an error, but
-                        // aggregate functions cannot presently produce errors.
-                        None
-                    } else {
-                        Some((key.unwrap_str(), val))
-                    }
-                })
-                .sorted_by_key(|(k, _v)| *k)
-                .dedup_by(|(k1, _v1), (k2, _v2)| k1 == k2),
-        );
+        let mut datums: Vec<_> = datums
+            .into_iter()
+            .filter_map(|d| {
+                if d.is_null() {
+                    return None;
+                }
+                let mut list = d.unwrap_list().iter();
+                let key = list.next().unwrap();
+                let val = list.next().unwrap();
+                if key.is_null() {
+                    // TODO(benesch): this should produce an error, but
+                    // aggregate functions cannot presently produce errors.
+                    None
+                } else {
+                    Some((key.unwrap_str(), val))
+                }
+            })
+            .collect();
+        // datums are ordered by any ORDER BY clause now, and we want to preserve the
+        // last entry for each key, but we also need to present unique and sorted keys
+        // to present_dict. Use sort_by here, which is stable, and so will preserve the
+        // ORDER BY order. Then reverse and dedup to retain the last of each
+        // key. Reverse again so we're back in push_dict order.
+        datums.sort_by_key(|(k, _v)| *k);
+        datums.reverse();
+        datums.dedup_by_key(|(k, _v)| *k);
+        datums.reverse();
+        packer.push_dict(datums);
     })
 }
 
-fn array_concat<'a, I>(datums: I, temp_storage: &'a RowArena) -> Datum<'a>
+fn order_aggregate_datums<'a, I>(
```

Comment:
I'm not sure if `array_concat` has vanished, but if this method is not exactly that any more, it would be great to get a doc comment explaining what it is and does. (( when its "I am the only definition of some SQL method", it seems less important ))